Facet and Filters behavior (#27)
meilisearch/specifications: text/0027-filter-and-facet-behavior.md

##### MeiliSearch v0.20.0 with `facetFilters`

In MeiliSeach v0.20.0, with the following documents

```json
[
  { "id": 456,  "genre": "adventure", "price": "12"  },
  { "id": 1,    "genre": "fantasy",   "price": "456" }
]
```
(the test is done with `price` as strings because MeiliSearch v0.20.0 cannot facet on numbers)

...and the following search

```json
{
    "q": "",
    "facetFilters": ["price:12"],
    "facetsDistribution": ["genre"]
}
```

...we get the following results:

```json
{
    "hits": [{ "id": 456,  "genre": "adventure", "price": "12" }],
    "facetsDistribution": {
        "genre": {
            "fantasy": 0,
            "adventure": 1
        }
    }
}
```

=> `fantasy` is set to `0` because the fantasy book does not have a `price` equals to `12` as required in the `facetFilters`.
=> the `filters` and the `facetsDistribution` are related: the `facetsDistribution` is applied after the facet filters.

##### Final decision for v0.21.0

MeiliSearch v0.21.0 will behave as the `facetFilters` already behave in MeiliSearch v0.20.0: the `facetsDistribution` will be applied after the filters.

> #### TLDR; all the breaking changes

Reviewer: @bidoubiwa @react-learner Do we need more precise indications for the documentation impact? Thanks!

Reviewer 2: Maybe juste some more information about the processing of the filters and facets. Is it done both at indexation time now? Or filters are still sorted at runtime?

Reviewer 2: Could we also have an example of an extreme filtering query? With a lot of facets, and filter, with AND and OR's? Nested in parentheses, etc

Reviewer: @bidoubiwa I've added some information. Hope it helps!

### III. Additional Materials

N/A

### IV. Explanation

#### Remove `facetFilters`, keep `filters`

The usage of `facetsFilter` is not needed anymore since everything is doable by only using the `filters` parameter.

```json
// Settings
{
    "attributesForFaceting": ["author"]
}
// Search
{
    "q": "",
    "filters": "price < 20",
    "facetFilters": ["author:'JK Rowling'"]
}
```

becomes

```json
// Settings
{
    "attributesForFaceting": ["author", "price"]
}
// Search
{
    "q": "",
    "filters": "price < 20 AND author = 'JK Rowling'"
}
```

We decided to replace the `:` operator in favor of `=`.

#### Operator behavior during the search

During the search, the operators should behave as they already do in MeiliSearch v0.20.0.

Here is a quick reminder of the v0.20.0 operator behavior:

- `<`, `=<`, `>`, `=>` => only operates on numbers. MeiliSearch returns only the documents with numbers for this field.
ex: `price > 19` does not return `"price": "20"` but returns `"price": 20` -> no document conversions are done.
- `=`, `!=`/`NOT` => string and number operation. MeiliSearch returns the documents with numbers, string (and array of strings) for this field.

> - cannot filter on `null`, objects, arrays of "undefined elements" (ex: array of `null`)

Reviewer: Which part? The one about returning documents?

> - cannot filter on `null`, objects, arrays of "undefined elements" (ex: array of `null`)

Reviewer 2: I think the limitied filtering may be linked to this behavior in meilisearch: https://docs.meilisearch.com/reference/under_the_hood/datatypes.html

> - cannot filter on `null`, objects, arrays of "undefined elements" (ex: array of `null`)

Reviewer 3: I meant 
```
- cannot filter on `null`, objects, arrays of "undefined elements" (ex: array of `null`)
```

> - cannot filter on `null`, objects, arrays of "undefined elements" (ex: array of `null`)

Reviewer: I've separated that from the operators behavior part for clarity !

```json
{
    "hits": [{ "id": 456,  "genre": "adventure", "price": "12" }],
    "facetsDistribution": {
        "genre": {
            "fantasy": 0,
            "adventure": 1
        }
    }
}
```

=> `fantasy` is set to `0` because the fantasy book does not have a `price` equals to `12` as required in the `facetFilters`.
=> the `filters` and the `facetsDistribution` are related: the `facetsDistribution` is applied after the facet filters.

##### Final decision for v0.21.0

MeiliSearch v0.21.0 will behave as the `facetFilters` already behave in MeiliSearch v0.20.0: the `facetsDistribution` will be applied after the filters.

#### TLDR; all the breaking changes

Here is the summary of all the breaking changes (that are detailed in the paragraphs above):

- The `facetFilters` parameter during the search is removed. Only `filters` can be used.
- The users need to set the attributes to `attributesForFaceting` to use the filters during the search via the `filters` parameters.
- The users can now pass an attribute containing numbers (float or integer) in `attributesForFaceting`. It means they can use `filters` and `facetsDistribution` on this numeric field.

> - The `filters` parameter can accept both syntax: string (with `OR`/`AND`) and array.

Reviewer 4: ```suggestion
- The `filters` parameter can accept both syntax: string (with `OR`/`AND`/`NOT`) and array.
```

### V. Impact on Documentation

See the previous part.

The documentation should present a complex filter query with multiple levels to precise that MeiliSearch is not limited to deepness level.

Example: 

```json
{
    "filters": "((genres = Comedy OR genres = Romance) AND (director = 'Mati Diop' OR director = 'Wong Kar-wai')) OR genres = 'Fantasy'"
}
```

## 2. Technical Aspects

### I. Abstract

Internal MeiliSearch engine uses an index data structure for each type encountered in an attribute. Doing this eliminate the need for the user to strictly type the attribute and permits to deal with loosely typed document at indexing time. Thus, facilitating the developper experience.

As for example, let’s imagine that we want to index two documents with a price attribute. One containing `"price": "20"` and the second `"price": 20` as value. The two attribute values will be stored in two different indexes.

Using `=` or `!=` operator on `price` attribute will lead the engine to query the two indexes and get matching document ids with an `UNION` operation. So it will return documents matching `"20"` or `20` as values for the `price` attribute.

#### Implementation Details

> We have a database for the facets, the keys are prefixed by the field_id (u8), a level (u8) and, the facet value (i64, f64). The facet values don't have a level when those are strings. The data stored under those keys is the document ids that are faceted under those facets values.

Reviewer 4: ```suggestion
We have a database for the facets, the keys are prefixed by the field_id (u8), a level (u8) and, the facet value (f64). The facet values don't have a level when the type is a string. The data stored under those keys is the document ids that are faceted under those facets values.
```

> The type of the facet (i.e. i64, f64 or string) is stored in another data structure and this is by using it that we know how to read the facet value. If the facet type is a number we are able to use more operators like greater than or lower than (e.g. <, <=, >, >=, =, !=).

Reviewer 4: ```suggestion
The type of the facet (i.e. f64 or string) is stored in another data structure and this is by using it that we know how to read the facet value. If the facet type is a number we are able to use more operators like greater than or lower than (e.g. <, <=, >, >=, =, !=).
```

##### Indexing phase

When documents come in and fields are declared as facets, we start storing the facet values in the previously described database, the key becomes the facet value (as a globally ordered byte slice) and, the entry data now contains the document id that contains this facet value. Note that if the facet value is a number we store it like [field id][level][left facet value][right facet value] where the level is 0 and if it is a string then we don't store the level.

Once the facet values that are numbers are stored we got a list of facet values prefixed with the field id and the base level (i.e. 0). We use this base level to generate more levels, each level contains groups of 4 groups of the level below, so level 1 aggregates the ids of the documents of each group of 4 facet values of level 0. The left and right facet values are the inclusive bounds of the group, the level 0 group have equal left and right bounds.

##### Querying phase

> Those levels are used to reduce the number of entries to run through, reducing the time it takes to answer too wide range filter queries, like duration > 0 where 80% of the entries will match. We go through each of the levels going from the higher one, the one which describes the biggest amount facet values and, we go deeper in the levels to find a better fitting bound.

Reviewer 4: ```suggestion
Those levels are used to reduce the number of entries to run through, reducing the time it takes to answer too wide range filter queries, like duration > 0 where 80% of the entries will match. We go through each of the levels going from the higher one, the one which describes the biggest amount of facet values and, we go deeper in the levels to find better fitting bounds.
```

### II. Issues Summary

> TBD

Author: ```suggestion

In Milli:
- https://github.com/meilisearch/milli/issues/152: the main issue gathering all the internal changes

In Transplant:
- https://github.com/meilisearch/transplant/issues/140: expose the both syntaxes for `filters`
- https://github.com/meilisearch/transplant/issues/70: the push of `attributesForFaceting`. The behavior does not change between v0.20.0 and v0.21.0 but it still needs changes in Transplant.
- https://github.com/meilisearch/transplant/issues/81: only keeping `filters` and removing `facetFilters`
```

Author: @gmourier

{
    "q": "",
    "filters": "price < 20",
    "facetFilters": ["author:'JK Rowling'"]
}
```

becomes

```json
// Settings
{
    "attributesForFaceting": ["author", "price"]
}
// Search
{
    "q": "",
    "filters": "price < 20 AND author = 'JK Rowling'"
}
```

We decided to replace the `:` operator in favor of `=`.

#### Operator behavior during the search

During the search, the operators should behave as they already do in MeiliSearch v0.20.0.

Here is a quick reminder of the v0.20.0 operator behavior:

- `<`, `=<`, `>`, `=>` => only operates on numbers. MeiliSearch returns only the documents with numbers for this field.
ex: `price > 19` does not return `"price": "20"` but returns `"price": 20` -> no field value conversions is done.
- `=`, `!=`/`NOT` => string and number operation. MeiliSearch returns the documents with numbers, string (and array of strings) for this field.
- cannot filter on `null`, objects, arrays of "undefined elements" (ex: array of `null`)

#### Accepted syntaxes for `filters`

Two syntaxes will be accepted for the `filters` parameter during the search:

- the string syntax with the `AND`/`OR`/`NOT` operators combined with parentheses
ex:
```json
{

>     "filters": "(genres = Comedy OR genres = Romance) AND director = 'Mati Diop'"

Reviewer: I've added a text to point that out. Nice catch!

}
```

> - the array syntax:

Author: @gmourier
Checked with @Kerollmops, one question: can we accept the string syntax in an element of the array syntax? Kero says it's technically possible 🙂 ex: `filters: [["price = 20 OR size < 12"], "NOT genre = comedy"]`

Reviewer: > I don't know if it's too late to suggest this, but I would strongly recommend we **change the `filters` parameter to just `filter`**. I think it's less confusing, mainly because the parameter's value will only ever be a single filter string or array.
> 
> Even if the value can be nested allowing for complexity, it's still just a logical expression. It's not as though you can define multiple filters that are applied independently or sequentially.
> 
> I would also recommend `filterDistribution` and `filterAttributes` to replace `facetsDistribution` and `attributesForFaceting`.

@curquiza Can we rename `filters` to `filter` at this time? I also think it helps to use it this way as a contextual verb.

@react-learner I think we need to keep `facetsDistribution` as is because this field is used to return computed facets. The parameter name is contextually ok. On another side, I agree with you on `attributesForFaceting` since it's now used to enable filtering on those listed fields and also to ask for MeiliSearch to expose those fields in `facetsDistribution`.

ex:
```json
{

>     "filters": [["genres = Comedy", "genres = Romance"], "director = 'Mati Diop'"]

Reviewer: You can express it like that `"filters": [["genres = Comedy", "director = 'Mati Diop'"], ["genres = Romance", "director = 'Mati Diop'"]]` since `(A && B) || C` is equal to `(A || C) && (B || C)`.

>     "filters": [["genres = Comedy", "genres = Romance"], "director = 'Mati Diop'"]

Reviewer 3: I see, thanks for explaining that @gmourier 😄

>     "filters": [["genres = Comedy", "genres = Romance"], "director = 'Mati Diop'"]

Reviewer: Thanks to @MarinPostma and @Kerollmops for helping me on this one 😅

=> the `filters` and the `facetsDistribution` are not related: the `facetsDistribution` is applied before the filters.

##### MeiliSearch v0.20.0 with `facetFilters`

In MeiliSeach v0.20.0, with the following documents

```json
[
  { "id": 456,  "genre": "adventure", "price": "12"  },
  { "id": 1,    "genre": "fantasy",   "price": "456" }
]
```
(the test is done with `price` as strings because MeiliSearch v0.20.0 cannot facet on numbers)

...and the following search

```json
{
    "q": "",
    "facetFilters": ["price:12"],
    "facetsDistribution": ["genre"]
}
```

...we get the following results:

```json
{
    "hits": [{ "id": 456,  "genre": "adventure", "price": "12" }],
    "facetsDistribution": {
        "genre": {
            "fantasy": 0,
            "adventure": 1
        }
    }
}
```

=> `fantasy` is set to `0` because the fantasy book does not have a `price` equals to `12` as required in the `facetFilters`.
=> the `filters` and the `facetsDistribution` are related: the `facetsDistribution` is applied after the facet filters.

##### Final decision for v0.21.0

> MeiliSearch v0.21.0 will behave as the `facetFilters` already behave in MeiliSearch v0.20.0: the `facetsDistribution` will be applied after the filters.

Reviewer 3: ```suggestion
In MeiliSearch v0.21.0, `facetsDistribution` will behave with `filters` the same way it currently does with `facetFilters`: the `facetsDistribution` will be applied after the filters.
```

As for example, let’s imagine that we want to index two documents with a price attribute. One containing `"price": "20"` and the second `"price": 20` as value. The two attribute values will be stored in two different indexes.

Using `=` or `!=` operator on `price` attribute will lead the engine to query the two indexes and get matching document ids with an `UNION` operation. So it will return documents matching `"20"` or `20` as values for the `price` attribute.

#### Implementation Details

We have a database for the facets, the keys are prefixed by the field_id (u8), a level (u8) and, the facet value (f64). The facet values don't have a level when the type is a string. The data stored under those keys is the document ids that are faceted under those facets values.

The type of the facet (i.e. f64 or string) is stored in another data structure and this is by using it that we know how to read the facet value. If the facet type is a number we are able to use more operators like greater than or lower than (e.g. <, <=, >, >=, =, !=).

##### Indexing phase

> When documents come in and fields are declared as facets, we start storing the facet values in the previously described database, the key becomes the facet value (as a globally ordered byte slice) and, the entry data now contains the document id that contains this facet value. Note that if the facet value is a number we store it like [field id][level][left facet value][right facet value] where the level is 0 and if it is a string then we don't store the level.

Reviewer 3: Make sure to put brackets in a `[code block]` so that MarkDown doesn't think you're trying to make a [link].

```suggestion
When documents come in and fields are declared as facets, we start storing the facet values in the previously described database, the key becomes the facet value (as a globally ordered byte slice) and, the entry data now contains the document id that contains this facet value. Note that if the facet value is a number we store it like `[field id][level][left facet value][right facet value]` where the level is 0 and if it is a string then we don't store the level.
```

ex: `price > 19` does not return `"price": "20"` but returns `"price": 20` -> no type conversions are done.
- `=`, `!=`/`NOT` => operate on string and number values. MeiliSearch returns only documents that have numbers, strings, or arrays of strings in this field.

#### Known limitations

- cannot filter on `null`, objects, arrays of "undefined elements" (ex: array of `null`)

#### Accepted syntaxes for `filter`

Three syntaxes will be accepted for the `filter` parameter during search. `String syntax`, `Array syntax` and `Mixed syntax`.

##### String syntax

The string syntax uses the `AND`/`OR`/`NOT` operators combined with parentheses to express a search filter.

Example:
```json
{
    "filter": "(genres = Comedy OR genres = Romance) AND director = 'Mati Diop'"
}
```

##### Array syntax

The array syntax uses dimensional array to express logical connectives.

- Inner arrays elements are connected by an OR operator (e.g. [["genres:Comedy", "genres:Romance"]]).
- Outer arrays elements are connected by an AND operator (e.g. ["genres:Romance", "director:Mati Diop"]).

Example:
```json
{
    "filter": [["genres = Comedy", "genres = Romance"], "director = 'Mati Diop'"]
}
```

> > The array syntax is limited to express a

Author: the sentence looks like it's not over

]
```
(the test is done with `price` as strings because MeiliSearch v0.20.0 cannot facet on numbers)

...and the following search

```json
{
    "q": "",
    "facetFilters": ["price:12"],
    "facetsDistribution": ["genre"]
}
```

...we get the following results:

```json
{
    "hits": [{ "id": 456,  "genre": "adventure", "price": "12" }],
    "facetsDistribution": {
        "genre": {
            "fantasy": 0,
            "adventure": 1
        }
    }
}
```

=> `fantasy` is set to `0` because the fantasy book does not have a `price` equals to `12` as required in the `facetFilters`.
=> the `filters` and the `facetsDistribution` are related: the `facetsDistribution` is applied after the facet filters.

##### Final decision for v0.21.0

In MeiliSearch v0.21.0, `facetsDistribution` will behave with `filter` the same way it currently does with `facetFilters`: the `facetsDistribution` will be applied after the filters.

#### TLDR; all the breaking changes

Here is the summary of all the breaking changes (that are detailed in the paragraphs above):

- The `facetFilters` parameter during the search is removed. Only `filter` can be used.

> - The `filters`parameter is renamed `filter`.

Author: ```suggestion
- The `filters` parameter is renamed `filter`.
```

- The `filters` parameter can accept both syntax: string (with `OR`/`AND`) and array.
- The `:` operator does not exist anymore (was previously present in `facetFilters` in v0.20.0) and is replaced by `=`.
- The `facetsDistribution` is now applied after the `filters` parameter. This point is currently not documented, not sure this is useful to add it to the docs.
- All the integer in the user documents are converted into float. So integers with high values lose precision. However, integers from −2^53 to 2^53 (−9007199254740992 to 9007199254740992) can be exactly represented, which is enough in 99% of cases. Not sure this is important to documented it either.

### Impact on Documentation

See the previous part.

## 2. Technical Specifications

> ### Architecture

Reviewer: Kerollmops I'm not quite 100% sure about this part. Don't hesitate to point out any errors. Thanks!

> 

Reviewer: Kerollmops I'm not quite 100% sure about the Architecture part. Don't hesitate to point out any errors. Thanks!

> Internal MeiliSearch engine uses an index data structure for each type encountered in an attribute. Doing this eliminate the need for the user to strictly type the attribute and permits to deal with loosely typed document at indexation. Thus, facilitating the developper experience.

Reviewer: @Kerollmops I'm not quite 100% sure about this part. Don't hesitate to point out any errors. Thanks!
